Metadata 2

xoviat opened the pull request:
Closes #45.

xoviat committed:
[flang2/ll_structure] add linker options to enum

xoviat committed:
[flang2/ll_structure] add psuedocode

xoviat committed:
[flang2/ll_write] impl. metadata name

xoviat committed:
[flang2/ll_structure] prelim. impl.

Probably wrong.

xoviat committed:
[flang2/ll_structure] cleanup

xoviat committed:
[flang2/ll_structure] fix type

xoviat committed:
[flang2/ll_structure] fix syntax

xoviat committed:
[flang2/ll_structure] fix type

xoviat committed:
fix

isuruf committed:
Update ll_structure.c

xoviat committed:
[flang2/ll_structure] cleanup

isuruf commented on tools/flang2/flang2exe/ll_structure.c:
```diff
@@ -578,6 +578,20 @@ ll_create_module(const char *module_name, const char *target_triple,
+     "/DEFAULTLIB:flang.lib",
+     "/DEFAULTLIB:flangrti.lib",
+     "/DEFAULTLIB:ompstub.lib"
+  };
```
I don't think we should code them here. There are various options sent to the driver like `-no-flang-libs`, `-static-flang-libs` that affect what should be added here.

xoviat replied:
Should we use a define or what?

isuruf replied:
I was thinking of an option to `flang2.exe`

xoviat replied:
Okay. Another problem is that the LLVM output is not correct. It should be like the debug option, but it's different.

isuruf commented:
```
C:\Users\idf2\AppData\Local\Temp\hello_world-a7edfe.ll:86:28: error: expected integer
!llvm.linker.options = !{ !"/DEFAULTLIB:flang.lib", !"/DEFAULTLIB:flangrti.lib", !"/DEFAULTLIB:ompstub.lib" }
```

xoviat committed:
fix type

xoviat committed:
fix LLVM IR

xoviat committed:
include header

ghost commented:
@isuruf Do you know how to get the command line arguments here? Generally in command line programs the arguments are parsed near the entry point.

isuruf commented:
Yes, working on it

xoviat committed:
move up header

xoviat committed:
[appveyor] cache build directory

xoviat committed:
Update .appveyor.yml

isuruf committed:
Add new options for linker_directives and llvm_target_triples

isuruf commented on tools/shared/utils/global.h:
```diff
@@ -159,6 +159,8 @@ typedef struct {
   LOGICAL ucase;
   char **idir;
+  char **linker_directives;
+  char *llvm_target_triple;
```
I was working on adding `llvm_target_triple`, so I combined that work. I can remove this and send a new PR. Let me know

isuruf commented on tools/flang2/flang2exe/ll_write.c:
```diff
@@ -1308,6 +1308,8 @@ get_metadata_name(enum LL_MDName name)
   case MD_llvm_dbg_cu:
     return "!llvm.dbg.cu";
+  case MD_llvm_linker_options:
+    return "!llvm.linker.options";
```
Is this supported in llvm 3.9 as well? Looks like https://reviews.llvm.org/D31349 is pretty new

xoviat replied:
Nope. Silly me.

isuruf replied:
That's fine. If it works for 4.0 then that's fine with me. We can work on 3.9 later

xoviat replied:
It doesn't work for 4.0. It works for 5.0. For 4.0, you have this:

```
!0 = !{i32 6, !"Linker Options", !1}
!1 = !{!2, !3}
!2 = !{!"/DEFAULTLIB:libcmtd.lib"}
!3 = !{!"/DEFAULTLIB:oldnames.lib"}
```

isuruf replied:
Shall I have another flag for `flang2.exe` for llvm version?

xoviat committed:
include header

isuruf committed:
Fix typo

xoviat committed:
fix header

isuruf committed:
Fix typo in llvm_target_triple

xoviat commented on tools/flang2/flang2exe/ll_structure.c:
```diff
@@ -578,6 +579,21 @@ ll_create_module(const char *module_name, const char *target_triple,
+    }
+    LL_MDRef linker_md = llmd_finish(mdb);
+    ll_extend_named_md_node(new_module, MD_llvm_linker_options, linker_md);
+  }
```
For 4.0, something like this:

```
LLMD_Builder mdb = llmd_init(new_module);
for (i = 0; (linker_directive = flg.linker_directives[i]); ++i) {
    LLMD_Builder submdb = llmd_init(new_module);
    llmd_add_string(submdb, linker_directive);
    LL_MDRef submd = llmd_finish(submdb);
    llmd_add_md(mdb, submd);
}
LL_MDRef md = llmd_finish(mdb);

LLMD_Builder boilerplatemdb = llmd_init(new_module);
llmd_add_i32(boilerplatemdb, 6);
llmd_add_string(boilerplatemdb, "Linker Options");
llmd_add_md(boilerplatemdb, md);
LL_MDRef boilerplatemd = llmd_finish(boilerplatemdb);

ll_extend_named_md_node(new_module, MD_llvm_module_flags, linker_md);
```

What a load of nonsense!

isuruf replied:
I'm getting `warning: ignoring debug info with an invalid version (0) in C:\Users\idf2\AppData\Local\Temp\hello_world-445182.ll`

xoviat replied:
What's the LLVM generated?

xoviat replied:
`ps: while ($true) { cp "$env:TMP\*.ll" C:\Downloads }`

isuruf replied:
https://gist.github.com/isuruf/d073e3b93552fde4565c293a10dd166e

xoviat replied:
`lldbg_init` will initialize the debug info (inc. the version). but are there other side effects?

xoviat committed:
move code to function

xoviat committed:
add defn to header

xoviat committed:
fix typo

isuruf committed:
Fix ll_structure.c

xoviat committed:
add debug version

ghost commented:
@isuruf Don't discuss on outdated changes please.

xoviat committed:
Update .appveyor.yml

isuruf commented:
@xoviat, @albertziegenhagel, when incorporating this in flang driver, should this be added for all platforms or just windows?

ghost commented:
Probably all platforms, but gcc has different arguments; these don't apply.

isuruf commented:
I mean will the following work for linux?
```
!llvm.module.flags = {!0}

!0 = !{i32 6, !"Linker Options", !1}
!1 = !{!2, !3}
!2 = !{!"-lflang"}
!3 = !{!"-lflangrti"}
```

xoviat committed:
fix

ghost commented:
Not sure. Also, this implementation as written could crash if `-g` is given (debug mode). Can you check that?

xoviat committed:
fix

xoviat committed:
fix

xoviat committed:
fix

isuruf commented:
I'll try this and get back to you

ghost commented:
Actually, what's stopping us from using just LLVM 5.0 (aside from ABI compatibility with older Pythons, which I don't really care about)?

isuruf commented:
Only the fact that the `llvmdev` package in `conda-forge` is not yet updated to 5.0

ghost commented:
https://anaconda.org/numba/llvmdev/files

isuruf committed:
ifdef for LLVM version

isuruf committed:
Fix for -linker -directive

When the linker directive starts with "-", flang2 interprets it as another option,
so flang.exe needs to send prepended with a \

isuruf commented on tools/flang2/flang2exe/ll_structure.c:
```diff
@@ -591,11 +591,14 @@ ll_create_module(const char *module_name, const char *target_triple,
 
 void
 add_linker_directives(LLVMModuleRef module) {
+#if LLVM_VERSION_MAJOR < 5
```
Instead of using the LLVM version like this, maybe we should send the version from `flang` so that it is checked at runtime.

xoviat replied:
I would personally prefer not to support LLVM < 5.0. Initializing the debug info here is questionable and not sure what the full implications are. But not sure sending the command line flags is a good idea, because we're building against LLVM anyway (I'd prefer to keep the binary "clean" in the newer version).

isuruf replied:
> because we're building against LLVM anyway

No, we are not. Flang is independent of the LLVM version. I just found out that `flang.exe` sends the LLVM version to `flang2.exe` and some parts of code use that. I've update the PR.

isuruf commented on tools/flang2/flang2exe/ll_structure.c:
```diff
@@ -591,11 +591,14 @@ ll_create_module(const char *module_name, const char *target_triple,
        LLMD_Builder submdb = llmd_init(module);
-
+       if (strlen(linker_directive) > 2 && linker_directive[0] == '\\' &&
+           linker_directive[1] == '-')
```
Major hack to support, `-linker -defaultlib:flangrti`. Basically `flang.exe` sends `-linker \-defaultlib:flangrti`

xoviat replied:
The way `cl.exe` does this is that all arguments after `-link` are simply passed down to the linker. So in the same way, we could just have a `-link` and then everything after it is just passed directly to the linker.

xoviat replied:
What do you think?

isuruf replied:
I don't want to make a big change like that without talking with upstream.

xoviat replied:
Let's talk now. Over to slack.

isuruf replied:
I reverted this change for now

isuruf committed:
Revert "Fix for -linker -directive"

This reverts commit 5deb6bf3ce432a1a2880d81c733ff6379e037e45.

isuruf committed:
Check LLVM version at runtime

xoviat committed:
[flang2:ll_structure] fix name
